### PR TITLE
DHCP Client now remembers previous IP

### PR DESCRIPTION
**Pull request by mipo57**
1. Fixed DHCP client sending ACK instead of REQUEST
2. Added IP retention - it is no longer changing on every reboot

### CODE REVIEW

File: NetworkControl/DHCPClientImplementation.cpp [hunk 1]
```diff
@@ -18,16 +18,21 @@ namespace Plugin {
...
         , _interfaceName(interfaceName)
         , _state(IDLE)
         , _xid(0)
         , _preferred()
+        , _last()
         , _callback(callback)
         , _offers()
+        , _persistentStorage(persistentStoragePath)
     {
+        _adminLock.Lock();
```
Comment (pwielders) [inline NetworkControl/DHCPClientImplementation.cpp]: During construction, I tjhink the risk of multiple concurency on this object is low. No need for locking here

File: NetworkControl/DHCPClientImplementation.cpp [hunk 2]
```diff
@@ -63,5 +68,55 @@ namespace Plugin {
...
+
+        return result;
+    }
+
+    void DHCPClientImplementation::SaveLastIP(const string& last_ip) 
+    {
+        if (!_persistentStorage.empty()) {
+            Core::Directory directory(_persistentStorage.c_str());
+
+            if (directory.CreatePath()) {
```
Comment (pwielders) [inline NetworkControl/DHCPClientImplementation.cpp]: As this is a one time action (creating the directory) if needed lets do this in the constructor of this object. Now there is a continous check on eachSaveLastIp()

File: NetworkControl/DHCPClientImplementation.cpp [hunk 3]
```diff
@@ -63,5 +68,55 @@ namespace Plugin {
...
+                Core::File file(_persistentStorage + lastIPFileName);
+                printf("SAVE FILE PATHNAME: %s", file.PathName().c_str());
```
Comment (pwielders) [inline NetworkControl/DHCPClientImplementation.cpp]: remove printf's the cluttter up the output with unwnated messages (in Release as well)

File: NetworkControl/DHCPClientImplementation.cpp [hunk 4]
```diff
@@ -63,5 +68,55 @@ namespace Plugin {
...
+                if (file.Exists()) {
+                    file.Open(true);
+                    file.SetSize(0);
+                } else {
+                    file.Create();
```
Comment (pwielders) [inline NetworkControl/DHCPClientImplementation.cpp]: I think with Create you can also create a file overwrite..

File: NetworkControl/DHCPClientImplementation.h [hunk 5]
```diff
@@ -140,6 +143,18 @@ namespace Plugin {
         static constexpr uint16_t DefaultDHCPServerPort = 67;
         static constexpr uint16_t DefaultDHCPClientPort = 68;
 
+        class IPStorage : public Core::JSON::Container {
+        public:
+            IPStorage()
```
Comment (pwielders) [inline NetworkControl/DHCPClientImplementation.h]: Please specify the CopyConstructor and Assignment operator as deleted.

File: NetworkControl/DHCPClientImplementation.cpp (hunk 4 above)
Comment (mipo57) [reply NetworkControl/DHCPClientImplementation.cpp]: I checked and I think I would need to add a O_TRUNC flag to open() function inside Create, but I can only change access flags.

**Comment by pwielders** [comment]
Good to go for me !

**Pull request merged**